File: java/LUIS/prediction/quickstart.java
import com.microsoft.azure.cognitiveservices.language.luis.authoring.*;
import com.microsoft.azure.cognitiveservices.language.luis.authoring.models.*;
import com.microsoft.azure.cognitiveservices.language.luis.runtime.*;
import com.microsoft.azure.cognitiveservices.language.luis.runtime.models.*;

import java.io.*;
import java.lang.Object.*;
import java.time.format.DateTimeFormatter;  
import java.time.LocalDateTime; 
import java.util.*;
import java.net.*;

/* 
 * To collect the Maven libraries needed, enter this at the command prompt:
 * mvn clean dependency:copy-dependencies
 * 
 * To compile and run, enter the following at a command prompt:
 * javac Quickstart.java -cp .;lib\*
 * java -cp .;lib\* Quickstart
 * 
 * This presumes your libraries are stored in a folder named "lib"
 * directly under the current folder. If not, please adjust the
 * -classpath/-cp value accordingly.
 */

public class Quickstart {
	/* Configure the local environment:
	* Set the following environment variables on your local machine using the
	* appropriate method for your preferred shell (Bash, PowerShell, Command
	* Prompt, etc.).
	*
	* LUIS_AUTHORING_KEY
	* LUIS_AUTHORING_ENDPOINT
	* LUIS_RUNTIME_KEY
	* LUIS_RUNTIME_ENDPOINT
	*
	* If the environment variable is created after the application is launched in a console or with Visual
	* Studio, the shell (or Visual Studio) needs to be closed and reloaded to take the environment variable into account.
	*/
    private static String authoring_key = System.getenv("LUIS_AUTHORING_KEY");
	private static String s_authoring_endpoint = System.getenv("LUIS_AUTHORING_ENDPOINT");
    private static String runtime_key = System.getenv("LUIS_RUNTIME_KEY");
	private static String s_runtime_endpoint = System.getenv("LUIS_RUNTIME_ENDPOINT");

// Note EndpointAPI.fromString cannot parse a URL that begins with "https://", so we remove it.
	private static com.microsoft.azure.cognitiveservices.language.luis.authoring.EndpointAPI authoring_endpoint = com.microsoft.azure.cognitiveservices.language.luis.authoring.EndpointAPI.fromString(s_authoring_endpoint.replace ("https://", ""));
	private static com.microsoft.azure.cognitiveservices.language.luis.runtime.EndpointAPI runtime_endpoint = com.microsoft.azure.cognitiveservices.language.luis.runtime.EndpointAPI.fromString(s_runtime_endpoint.replace ("https://", ""));

	private class AppInfo {
		public String app_id;
		public String app_version;
		public AppInfo(String app_id, String app_version) {
			this.app_id = app_id;
			this.app_version = app_version;
		}
	}

	public static void list_applications() {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		Apps.AppsListDefinitionStages.WithExecute stages = client.apps().list();
		List<ApplicationInfoResponse> result = stages.execute();

		for (ApplicationInfoResponse app : result) { 
			System.out.print(app.id().toString()); 
		} 
	}

	public AppInfo create_application() {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		PrebuiltDomainCreateObject create_app_payload = new PrebuiltDomainCreateObject().withDomainName("HomeAutomation").withCulture("en-us");
		UUID app_id = client.apps().addCustomPrebuiltDomain(create_app_payload);
		System.out.println("Created LUIS app with ID " + app_id.toString());
		return new AppInfo(app_id.toString(), "0.1");
	}

	public void train_application(AppInfo app_info) throws InterruptedException {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		client.trains().trainVersion(UUID.fromString(app_info.app_id), app_info.app_version);

		// Wait for the train operation to finish.
		System.out.println ("Waiting for train operation to finish...");

		Boolean waiting = true;
		while (true == waiting) {
			List<ModelTrainingInfo> result = client.trains().getStatus(UUID.fromString(app_info.app_id), app_info.app_version);
			// GetStatus returns a list of training statuses, one for each model.
			// Loop through them and make sure all are done.
			waiting = false;
			for (ModelTrainingInfo info : result) { 
				String status = info.details().status();
				if (status.equals("Queued") || status.equals("InProgress")) {
					waiting = true;
				}
				if (status.equals("Fail")) {
					System.out.println("Training operation failed. Reason: " + info.details().failureReason());
				}
			}
			if (true == waiting) {
				System.out.println("Waiting 10 seconds for training to complete...");
				Thread.sleep(10000);
			}
		}
	}

	public void publish_application(AppInfo app_info) {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		/* NOTE: Make sure you use the same value for staging when you publish the app and when you query it.
		Otherwise, the query raises an exception that the app was not found. */
		 ProductionOrStagingEndpointInfo result = client.apps().publish(UUID.fromString(app_info.app_id), new ApplicationPublishObject().withVersionId(app_info.app_version).withIsStaging(true));
		
		System.out.println("Application published. Endpoint URL: " + result.endpointUrl());
	}

	public void query_application(AppInfo app_info) {
		LuisRuntimeAPI client = LuisRuntimeManager.authenticate(runtime_endpoint, runtime_key);
		String query = "turn on the lights";

		/* NOTE: Make sure you use the same value for staging when you publish the app and when you query it.
		Otherwise, the query raises an exception that the app was not found. */
		LuisResult result = client.predictions().resolve().withAppId(app_info.app_id).withQuery(query).withStaging(true).execute();
		System.out.println("Top scoring intent: " + result.topScoringIntent().intent());
		if (result.sentimentAnalysis() != null) {
			System.out.println("Sentiment label: " + result.sentimentAnalysis().label() + ". Score: " + result.sentimentAnalysis().score());
		}
		if (result.intents() != null) {
			System.out.println("Intents: ");
			for (IntentModel intent : result.intents()) { 
				System.out.println("\t" + intent.intent() + ". Score: " + intent.score());
			}
		}
		if (result.entities() != null) {
			System.out.println("Entities: ");
			for (EntityModel entity : result.entities()) { 
				System.out.println("\t" + entity.entity() + ". Type: " + entity.type());
			}
		}
	}

	public void delete_application(AppInfo app_info) {
		LUISAuthoringClient client = LUISAuthoringManager.authenticate(authoring_endpoint, authoring_key);

		OperationStatus status = client.apps().delete(UUID.fromString(app_info.app_id));
		OperationStatusType code = status.code();
		System.out.println("Deleted LUIS app with ID " + app_info.app_id + ". Operation status: " + code);
	}

    public static void main(String[] args) {
		try {
			Quickstart quickstart = new Quickstart();
			AppInfo app_info = quickstart.create_application();

			quickstart.train_application(app_info);
			quickstart.publish_application(app_info);
			quickstart.query_application(app_info);
			quickstart.delete_application(app_info);
        } catch (Exception e) {
            System.out.println(e.getMessage());
            e.printStackTrace();
        }
    }
}

File: javascript/LUIS/prediction/quickstart.js
"use strict";

/* To run this sample, install the following modules.
 * npm install @azure/cognitiveservices-luis-authoring
 * npm install @azure/cognitiveservices-luis-runtime
 * npm install @azure/ms-rest-js
 */
var authoring = require("@azure/cognitiveservices-luis-authoring");
var runtime = require("@azure/cognitiveservices-luis-runtime");
var msRest = require("@azure/ms-rest-js");

/* Configure the local environment:
* Set the following environment variables on your local machine using the
* appropriate method for your preferred shell (Bash, PowerShell, Command
* Prompt, etc.).
*
* LUIS_AUTHORING_KEY
* LUIS_AUTHORING_ENDPOINT
* LUIS_RUNTIME_KEY
* LUIS_RUNTIME_ENDPOINT
*
* If the environment variable is created after the application is launched in a console or with Visual
* Studio, the shell (or Visual Studio) needs to be closed and reloaded to take the environment variable into account.
*/
const authoring_key_var = 'LUIS_AUTHORING_KEY';
if (!process.env[authoring_key_var]) {
    throw new Error('please set/export the following environment variable: ' + authoring_key_var);
}
const authoring_key = process.env[authoring_key_var];

const authoring_endpoint_var = 'LUIS_AUTHORING_ENDPOINT';
if (!process.env[authoring_endpoint_var]) {
    throw new Error('please set/export the following environment variable: ' + authoring_endpoint_var);
}
const authoring_endpoint = process.env[authoring_endpoint_var];

const runtime_key_var = 'LUIS_RUNTIME_KEY';
if (!process.env[runtime_key_var]) {
    throw new Error('please set/export the following environment variable: ' + runtime_key_var);
}
const runtime_key = process.env[runtime_key_var];

const runtime_endpoint_var = 'LUIS_RUNTIME_ENDPOINT';
if (!process.env[runtime_endpoint_var]) {
    throw new Error('please set/export the following environment variable: ' + runtime_endpoint_var);
}
const runtime_endpoint = process.env[runtime_endpoint_var];

const authoring_creds = new msRest.ApiKeyCredentials({ inHeader: { 'Ocp-Apim-Subscription-Key': authoring_key } });
const authoring_client = new authoring.LUISAuthoringClient(authoring_creds, authoring_endpoint);

const runtime_creds = new msRest.ApiKeyCredentials({ inHeader: { 'Ocp-Apim-Subscription-Key': runtime_key } });
const runtime_client = new runtime.LUISRuntimeClient(runtime_creds, runtime_endpoint);

function create_app() {
    var create_app_payload = {
        domainName: "HomeAutomation",
        culture: "en-us"
    };
    return authoring_client.apps.addCustomPrebuiltDomain(create_app_payload).then((result) => {
        console.log("Created LUIS app with ID " + result.body);
        return new AppInfo(result.body, "0.1");
    }).catch(error => {
        throw error;
    });
}

function wait_for_operation(app_info) {
    return authoring_client.train.getStatus(app_info.id, app_info.version).then(function (result) {
        // GetStatus returns a list of training statuses, one for each model.
        // Loop through them and make sure all are done.
        if (result.some(function (info) {
            var status = info.details.status;
            if ("Queued" === status || "InProgress" === status) {
                return true;
            }
            else if ("Fail" === status) {
                console.log("Training operation failed. Reason: " + info.details.failureReason);
                return false;
            }
            else {
                return false;
            }
        })) {
            console.log("Waiting 10 seconds for training to complete...");
            return new Promise(function (resolve, reject) {
                setTimeout(function () {
                    resolve(wait_for_operation(app_info));
                }, 10000);
            });
        }
    }).catch(error => {
        throw error;
    });
}

function train_app(app_info) {
    return authoring_client.train.trainVersion(app_info.id, app_info.version).then((result) => {
        // Wait for the train operation to finish.
        console.log("Waiting for train operation to finish...");
        return wait_for_operation(app_info);
    }).catch(error => {
        throw error;
    });
}

function publish_app(app_info) {
    return authoring_client.apps.publish(app_info.id, { versionId: app_info.version, isStaging: true }).then((result) => {
        console.log("Application published. Endpoint URL: " + result.endpointUrl);    
    }).catch(error => {
        throw error;
    });
}

function predict(app_info) {
    var ops = new runtime.PredictionOperations(runtime_client);
// Note be sure to specify, using the slotName parameter, whether your application is in staging or production.
// By default, applications are created in staging.
	var slotName = "staging";
    var request = { query: "turn on the bedroom light" };
    return ops.getSlotPrediction(app_info.id, slotName, request).then((result) => {
        console.log(result);
    }).catch(error => {
        throw error;
    })
}

function delete_app(app_info) {
    return authoring_client.apps.deleteMethod(app_info.id).then((result) => {
        console.log("Application with ID " + app_info.id + " deleted. Operation result: " + result.message);
    }).catch(error => {
        throw error;
    })
}

async function quickstart() {
    var app_info = await create_app();
    await train_app(app_info);
    await publish_app(app_info);
	await predict(app_info);
    await delete_app(app_info);
}

try {
    quickstart();
}
catch (error) {
    console.log(error);
}
